### PR TITLE
Add toImage to RenderRepaintBoundary

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds a toImage function to RenderRepaintBoundary that returns an uncompressed raw image of the RenderRepaintBoundary and its children.  A device pixel ratio different from the physical ratio may be specified for the captured image.  A value of 1.0 will give an image in logical pixels.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -1103,6 +1104,45 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   /// Override in subclasses with children and call the visitor for each child.
   void visitChildren(RenderObjectVisitor visitor) { }
 
+  /// Capture an image of the current state of the render object and its
+  /// children.
+  ///
+  /// The returned [ui.Image] has uncompressed raw bytes in the dimensions of
+  /// the render object, multiplied by the [window.devicePixelRatio].
```
Comment (tvolkert) [inline packages/flutter/lib/src/rendering/object.dart]: Don't say "multiplied by the [window.devicePixelRatio]" since it's covered below in the discussion of the [pixelRatio] named parameter?

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -1103,6 +1104,45 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  ///
+  /// The property [isRepaintBoundary] must be true for [toImage] to work, and
+  /// the render object must not need layout, painting, or compositing.
```
Comment (tvolkert) [inline packages/flutter/lib/src/rendering/object.dart]: > the render object must not need layout, painting, or compositing

How does the caller know whether this is the case?

Also, can we assert it below?

File: packages/flutter/lib/src/rendering/object.dart [hunk 3]
```diff
@@ -1103,6 +1104,45 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  ///
+  /// The [pixelRatio] describes the scale between the logical pixels and the
+  /// size of the output image. It is independent of the
+  /// [window.devicePixelRatio] for the device, so specifying 1.0 (the default)
+  /// will give you a 1:1 mapping between logical pixels and the output pixels
+  /// in the image. The output image is not anti-aliased, so you may wish to
```
Comment (tvolkert) [inline packages/flutter/lib/src/rendering/object.dart]: Where is anti-aliasing usually done when the render object is drawn to the screen?

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -1103,6 +1104,45 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  /// use a larger device pixel ratio to obtain an image large enough to
+  /// down sample. If you wish to have an image the same size as the actual
+  /// device pixels, then specify [window.devicePixelRatio] as the [pixelRatio].
+  ///
+  /// See also:
+  ///
+  ///  * [Scene.toImage] for more information about the image returned.
```
Comment (tvolkert) [inline packages/flutter/lib/src/rendering/object.dart]: Does this have to be `[ui.Scene.toImage]`?

File: packages/flutter/lib/src/rendering/object.dart [hunk 5]
```diff
@@ -1103,6 +1104,45 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  Future<ui.Image> toImage({double pixelRatio: 1.0}) async {
+    assert(isRepaintBoundary, 'You can only call toImage for RenderObjects that are repaint boundaries.');
+    final ui.SceneBuilder builder = new ui.SceneBuilder();
+    Matrix4 transform = new Matrix4.diagonal3Values(pixelRatio, pixelRatio, 0.0);
+    transform = transform.multiplied(
```
Comment (tvolkert) [inline packages/flutter/lib/src/rendering/object.dart]: Is it worth optimizing out this matrix multiply for the case where `pixelRatio` is 1.0?

File: packages/flutter/lib/src/rendering/object.dart [hunk 6]
```diff
@@ -1103,6 +1104,45 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+      new Matrix4.translationValues(-layer.offset.dx, -layer.offset.dy, 0.0),
+    );
+    builder.pushTransform(transform.storage);
+    layer.addToScene(builder, Offset.zero);
+    final ui.Scene scene = builder.build();
+    final ui.Image result = await scene.toImage(
+      (pixelRatio * paintBounds.width).round(),
+      (pixelRatio * paintBounds.height).round(),
+    );
+    scene.dispose();
```
Comment (tvolkert) [inline packages/flutter/lib/src/rendering/object.dart]: `try/finally` with a call to `scene.dispose()` in the finally?

File: packages/flutter/test/rendering/object_test.dart [hunk 7]
```diff
@@ -23,16 +26,33 @@ void main() {
...
-    final TestRenderObject renderObject = new TestRenderObject();
+    final TestSemanticRenderObject renderObject = new TestSemanticRenderObject();
     expect(renderObject.attached, isFalse);
     expect(renderObject.describeSemanticsConfigurationCallCount, 0);
 
     renderObject.markNeedsSemanticsUpdate();
     expect(renderObject.describeSemanticsConfigurationCallCount, 0);
   });
+
+  test('RenderObject can capture images of itself', () async {
```
Comment (tvolkert) [inline packages/flutter/test/rendering/object_test.dart]: Add a test that expects our assertion error if the render object isn't a repaint boundary?

**Review by tvolkert — APPROVED** [review]
LGTM

**Review by tvolkert — APPROVED** [review]

File: packages/flutter/lib/src/rendering/object.dart [hunk 8]
```diff
@@ -1103,6 +1104,46 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  /// The [pixelRatio] describes the scale between the logical pixels and the
+  /// size of the output image. It is independent of the
+  /// [window.devicePixelRatio] for the device, so specifying 1.0 (the default)
+  /// will give you a 1:1 mapping between logical pixels and the output pixels
+  /// in the image.
+  ///
+  /// See also:
+  ///
+  ///  * [ui.Scene.toImage] for more information about the image returned.
+  Future<ui.Image> toImage({double pixelRatio: 1.0}) async {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: To avoid people depending on particular widgets that happen to be repaint boundaries but might not always be, I recommend putting this on RenderRepaintBoundary.

File: packages/flutter/lib/src/rendering/object.dart [hunk 9]
```diff
@@ -1103,6 +1104,46 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+    Matrix4 transform = new Matrix4.diagonal3Values(pixelRatio, pixelRatio, 1.0);
+    transform = transform.multiplied(
+      new Matrix4.translationValues(-layer.offset.dx, -layer.offset.dy, 0.0),
+    );
+    builder.pushTransform(transform.storage);
+    layer.addToScene(builder, Offset.zero);
+    final ui.Scene scene = builder.build();
+    try {
+      return await scene.toImage(
+        (pixelRatio * paintBounds.width).round(),
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: `paintBounds` is a bit of a dubious choice because it's just advisory really.
If you move this to RenderRepaintBoundary, you can just use `size`, which is much more authoritative.

File: packages/flutter/lib/src/rendering/object.dart [hunk 10]
```diff
@@ -1103,6 +1104,46 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   /// Override in subclasses with children and call the visitor for each child.
   void visitChildren(RenderObjectVisitor visitor) { }
 
+  /// Capture an image of the current state of the render object and its
+  /// children.
+  ///
+  /// The returned [ui.Image] has uncompressed raw bytes in the dimensions of
+  /// the render object, multiplied by the [pixelRatio].
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: We should replace "dimensions" with the actual property that we use to determine the dimensions (currently `[paintBounds]`, though if you take my feedback below, `[size]`).

File: packages/flutter/lib/src/rendering/object.dart [hunk 11]
```diff
@@ -1103,6 +1104,46 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  ///
+  /// The property [isRepaintBoundary] must be true for [toImage] to work, and
+  /// the render object must not need painting (i.e. [debugNeedsPaint] must be
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: rather than "must not need painting" it might be more positive to say "must have gone through the paint phase" or some such (the parenthetical is fine though)

File: packages/flutter/lib/src/rendering/object.dart [hunk 12]
```diff
@@ -1103,6 +1104,46 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+  ///
+  ///  * [ui.Scene.toImage] for more information about the image returned.
+  Future<ui.Image> toImage({double pixelRatio: 1.0}) async {
+    assert(isRepaintBoundary, 'You can only call toImage for RenderObjects that are repaint boundaries.');
+    assert(!_needsPaint);
+    final ui.SceneBuilder builder = new ui.SceneBuilder();
+    Matrix4 transform = new Matrix4.diagonal3Values(pixelRatio, pixelRatio, 1.0);
+    transform = transform.multiplied(
+      new Matrix4.translationValues(-layer.offset.dx, -layer.offset.dy, 0.0),
+    );
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: rather than using `multiplied()` and allocating a new matrix, you can just use `translate`.

File: packages/flutter/lib/src/rendering/object.dart [hunk 13]
```diff
@@ -1103,6 +1104,46 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
+    builder.pushTransform(transform.storage);
+    layer.addToScene(builder, Offset.zero);
+    final ui.Scene scene = builder.build();
+    try {
+      return await scene.toImage(
+        (pixelRatio * paintBounds.width).round(),
+        (pixelRatio * paintBounds.height).round(),
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: I wonder if we should change toImage to just take a Size... cc @tvolkert

File: packages/flutter/test/rendering/object_test.dart (hunk 7 above)
Comment (Hixie) [reply packages/flutter/test/rendering/object_test.dart]: we do test asserts sometimes (check for `takeException`), but this point will be moot anyway if you move it to RenderRepaintBoundary.

File: packages/flutter/test/rendering/object_test.dart [hunk 14]
```diff
@@ -23,16 +26,33 @@ void main() {
...
+    expect(image.height, equals(200));
+
+    // Now with pixel ratio set to something other than 1.0.
+    boundary = new RenderRepaintBoundary();
+    layout(boundary, constraints: new BoxConstraints.tight(const Size(100.0, 200.0)));
+    pumpFrame(phase: EnginePhase.composite);
+    image = await boundary.toImage(pixelRatio: 2.0);
+    expect(image.width, equals(200));
+    expect(image.height, equals(400));
+  });
```
Comment (Hixie) [inline packages/flutter/test/rendering/object_test.dart]: You should add a test that verifies that it works (gives the right pixels) with two layers, since that was a bug.

File: packages/flutter/lib/src/rendering/object.dart (hunk 10 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/object.dart]: OK, size it is.

File: packages/flutter/lib/src/rendering/object.dart (hunk 11 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/object.dart]: Done.

File: packages/flutter/lib/src/rendering/object.dart (hunk 8 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/object.dart]: Yeah, I guess that makes sense.  It means that they'll have to cast the result when they get the render object in order to call this function, but that's not horrible.

File: packages/flutter/lib/src/rendering/object.dart (hunk 9 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/object.dart]: Done.

File: packages/flutter/lib/src/rendering/object.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/object.dart]: I don't really understand how we're actually rendering this, but I would presume we're just uploading the whole scene to the GPU, having it render to a texture, then grabbing the texture? In which case I'd expect antialiasing to work fine?

cc @jason-simmons or @chinmaygarde

File: packages/flutter/lib/src/rendering/object.dart (hunk 13 above)
Comment (tvolkert) [reply packages/flutter/lib/src/rendering/object.dart]: For the use cases we've heard for this, I don't have a strong opinion either way.

File: packages/flutter/lib/src/rendering/object.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/object.dart]: search for `[ui.` or `[dart:ui.` and compare to actual output, I forget which one is the right one.

File: packages/flutter/lib/src/rendering/object.dart (hunk 12 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/object.dart]: Good idea.  Done.

File: packages/flutter/test/rendering/object_test.dart (hunk 14 above)
Comment (gspencergoog) [reply packages/flutter/test/rendering/object_test.dart]: Done.

**Comment by tvolkert** [comment]
I tried this out in a simple widget test:

```dart
testWidgets('test', (WidgetTester tester) async {
  await tester.pumpWidget(new MyApp());

  RenderRepaintBoundary repaintRenderBoundary = tester
      .allRenderObjects
      .firstWhere((RenderObject object) => object is RenderRepaintBoundary);
  ui.Image image = await repaintRenderBoundary.toImage();
  ByteData bytes = await image.toByteData(format: ui.ImageByteFormat.png);
});
```

If I run that test on an Android device/emulator, I get a correct PNG capture of the app.  However, if I run that on the host in `flutter_tester`, the PNG is all white.

Am I holding it wrong?

**Comment by tvolkert** [comment]
aha - indeed I _was_ holding it wrong.  I was calling:

```dart
await new File('/path/fo/foo.png').writeAsBytes(bytes);
```

Which doesn't play nicely with our fake async.  When I call `writeAsBytesSync(bytes)`, it all worked great.

File: packages/flutter/lib/src/rendering/object.dart (hunk 3 above)
Comment (chinmaygarde) [reply packages/flutter/lib/src/rendering/object.dart]: Yeah, what @Hixie said. We have to render to texture on the GPU first because the scene may reference images that are only resident on the GPU. AA should work just fine.
